tests/evaluate/test_default_evaluator.py

```diff
@@ -2990,6 +2982,135 @@ def test_default_metrics_as_custom_metrics_static_dataset():
     assert "exact_match/v1" in results.metrics.keys()
 
 
+def test_derived_metrics():
+    def metric_1(predictions, targets, metrics):
```

Comment:
@annzhang-db What happens if the user has a custom metric with a hyphen or slash in the name, etc? Do we throw? If not, we can't use a Python variable name to refer to it in derived metrics, right? Same problem for metrics depending on particular dataframe columns. Are we handling this already?

Reply:
Following up on this, I tried to run this example where I compute a derived metric `answer_similarity/v1`, but I encountered an unintended / internal exception:

```
import os

import openai
import pandas as pd

import mlflow
from mlflow.metrics import make_metric
from mlflow.metrics.genai import EvaluationExample, answer_similarity

assert "OPENAI_API_KEY" in os.environ, "Please set the OPENAI_API_KEY environment variable."


def custom_metric(targets, predictions, answer_similarity_v1):
    print("answer sim", answer_similarity_v1_mean)
    return 8 # Some dummy value


# testing with OpenAI gpt-3.5-turbo
example = EvaluationExample(
    input="What is MLflow?",
    output="MLflow is an open-source platform for managing machine "
    "learning workflows, including experiment tracking, model packaging, "
    "versioning, and deployment, simplifying the ML lifecycle.",
    score=4,
    justification="The definition effectively explains what MLflow is "
    "its purpose, and its developer. It could be more concise for a 5-score.",
    grading_context={
        "targets": "MLflow is an open-source platform for managing "
        "the end-to-end machine learning (ML) lifecycle. It was developed by Databricks, "
        "a company that specializes in big data and machine learning solutions. MLflow is "
        "designed to address the challenges that data scientists and machine learning "
        "engineers face when developing, training, and deploying machine learning models."
    },
)

answer_similarity_metric = answer_similarity(examples=[example])

eval_df = pd.DataFrame(
    {
        "inputs": [
            "What is MLflow?",
            "What is Spark?",
            "What is Python?",
        ],
        "ground_truth": [
            "MLflow is an open-source platform for managing the end-to-end machine learning (ML) lifecycle. It was developed by Databricks, a company that specializes in big data and machine learning solutions. MLflow is designed to address the challenges that data scientists and machine learning engineers face when developing, training, and deploying machine learning models.",
            "Apache Spark is an open-source, distributed computing system designed for big data processing and analytics. It was developed in response to limitations of the Hadoop MapReduce computing model, offering improvements in speed and ease of use. Spark provides libraries for various tasks such as data ingestion, processing, and analysis through its components like Spark SQL for structured data, Spark Streaming for real-time data processing, and MLlib for machine learning tasks",
            "Python is a high-level programming language that was created by Guido van Rossum and released in 1991. It emphasizes code readability and allows developers to express concepts in fewer lines of code than languages like C++ or Java. Python is used in various domains, including web development, scientific computing, data analysis, and machine learning.",
        ],
    }
)

with mlflow.start_run() as run:
    system_prompt = "Answer the following question in two sentences"
    logged_model = mlflow.openai.log_model(
        model="gpt-3.5-turbo",
        task=openai.ChatCompletion,
        artifact_path="model",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": "{question}"},
        ],
    )

    results = mlflow.evaluate(
        logged_model.model_uri,
        eval_df,
        targets="ground_truth",
        model_type="question-answering",
        extra_metrics=[
            answer_similarity_metric,
            make_metric(
                eval_fn=custom_metric,
                greater_is_better=False,
            )
        ],
    )
    print(results)

    eval_table = results.tables["eval_results_table"]
    print(eval_table)
```

It failed with

```
 $ python evaluate_with_llm_judge.py
/Users/corey.zumar/miniconda3/lib/python3.11/site-packages/pydantic/_internal/_config.py:321: UserWarning: Valid config keys have changed in V2:
* 'schema_extra' has been renamed to 'json_schema_extra'
  warnings.warn(message, UserWarning)
/Users/corey.zumar/mlflow/mlflow/data/digest_utils.py:26: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/01/30 23:40:57 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/01/30 23:40:57 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
Traceback (most recent call last):
  File "/Users/corey.zumar/mlflow/examples/evaluation/evaluate_with_llm_judge.py", line 68, in <module>
    results = mlflow.evaluate(
              ^^^^^^^^^^^^^^^^
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 1880, in evaluate
    evaluate_result = _evaluate(
                      ^^^^^^^^^^
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 1120, in _evaluate
    eval_result = evaluator.evaluate(
                  ^^^^^^^^^^^^^^^^^^^
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1914, in evaluate
    evaluation_result = self._evaluate(model, is_baseline_model=False)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1803, in _evaluate
    self._evaluate_metrics(eval_df)
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1596, in _evaluate_metrics
    self._order_extra_metrics(eval_df)
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1556, in _order_extra_metrics
    self._raise_exception_for_malformed_metrics(failed_results, eval_df)
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1517, in _raise_exception_for_malformed_metrics
    error_messages = [
                     ^
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 1519, in <listcomp>
    for metric_name, param_names in malformed_results
        ^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: not enough values to unpack (expected 2, got 1)

```

Looks like MLflow hits an error when trying to surface information about another error :)

Reply:
I then retried this by changing the `answer_similarity_v1` parameter in the metric function to `answer_similarity`. However, it seems like the value of `answer_similarity` is `None`, even though the builtin metric was computed (see the line `answer sim None` in the output):

```
import os

import openai
import pandas as pd

import mlflow
from mlflow.metrics import make_metric
from mlflow.metrics.genai import EvaluationExample, answer_similarity

assert "OPENAI_API_KEY" in os.environ, "Please set the OPENAI_API_KEY environment variable."


def custom_metric(targets, predictions, answer_similarity):
    print("answer sim", answer_similarity)
    return 8 # Some dummy value


# testing with OpenAI gpt-3.5-turbo
example = EvaluationExample(
    input="What is MLflow?",
    output="MLflow is an open-source platform for managing machine "
    "learning workflows, including experiment tracking, model packaging, "
    "versioning, and deployment, simplifying the ML lifecycle.",
    score=4,
    justification="The definition effectively explains what MLflow is "
    "its purpose, and its developer. It could be more concise for a 5-score.",
    grading_context={
        "targets": "MLflow is an open-source platform for managing "
        "the end-to-end machine learning (ML) lifecycle. It was developed by Databricks, "
        "a company that specializes in big data and machine learning solutions. MLflow is "
        "designed to address the challenges that data scientists and machine learning "
        "engineers face when developing, training, and deploying machine learning models."
    },
)

answer_similarity_metric = answer_similarity(examples=[example])

eval_df = pd.DataFrame(
    {
        "inputs": [
            "What is MLflow?",
            "What is Spark?",
            "What is Python?",
        ],
        "ground_truth": [
            "MLflow is an open-source platform for managing the end-to-end machine learning (ML) lifecycle. It was developed by Databricks, a company that specializes in big data and machine learning solutions. MLflow is designed to address the challenges that data scientists and machine learning engineers face when developing, training, and deploying machine learning models.",
            "Apache Spark is an open-source, distributed computing system designed for big data processing and analytics. It was developed in response to limitations of the Hadoop MapReduce computing model, offering improvements in speed and ease of use. Spark provides libraries for various tasks such as data ingestion, processing, and analysis through its components like Spark SQL for structured data, Spark Streaming for real-time data processing, and MLlib for machine learning tasks",
            "Python is a high-level programming language that was created by Guido van Rossum and released in 1991. It emphasizes code readability and allows developers to express concepts in fewer lines of code than languages like C++ or Java. Python is used in various domains, including web development, scientific computing, data analysis, and machine learning.",
        ],
    }
)

with mlflow.start_run() as run:
    system_prompt = "Answer the following question in two sentences"
    logged_model = mlflow.openai.log_model(
        model="gpt-3.5-turbo",
        task=openai.ChatCompletion,
        artifact_path="model",
        messages=[
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": "{question}"},
        ],
    )

    results = mlflow.evaluate(
        logged_model.model_uri,
        eval_df,
        targets="ground_truth",
        model_type="question-answering",
        extra_metrics=[
            answer_similarity_metric,
            make_metric(
                eval_fn=custom_metric,
                greater_is_better=False,
            )
        ],
    )
    print(results)

    eval_table = results.tables["eval_results_table"]
    print(eval_table)
```

```
$ python evaluate_with_llm_judge.py
/Users/corey.zumar/miniconda3/lib/python3.11/site-packages/pydantic/_internal/_config.py:321: UserWarning: Valid config keys have changed in V2:
* 'schema_extra' has been renamed to 'json_schema_extra'
  warnings.warn(message, UserWarning)
/Users/corey.zumar/mlflow/mlflow/data/digest_utils.py:26: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  string_columns = trimmed_df.columns[(df.applymap(type) == str).all(0)]
/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py:414: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  data = data.applymap(_hash_array_like_element_as_bytes)
2024/01/30 23:43:32 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
2024/01/30 23:43:32 INFO mlflow.models.evaluation.default_evaluator: Computing model predictions.
2024/01/30 23:43:34 INFO mlflow.models.evaluation.default_evaluator: Testing metrics on first row...
2024/01/30 23:43:34 WARNING mlflow.metrics.metric_definitions: Failed to load 'toxicity' metric (error: ModuleNotFoundError("No module named 'evaluate'")), skipping metric logging.
2024/01/30 23:43:34 WARNING mlflow.models.evaluation.default_evaluator: Did not log builtin metric 'toxicity' because it returned None.
2024/01/30 23:43:34 WARNING mlflow.metrics.metric_definitions: Failed to load flesch kincaid metric, skipping metric logging.
2024/01/30 23:43:34 WARNING mlflow.models.evaluation.default_evaluator: Did not log builtin metric 'flesch_kincaid_grade_level' because it returned None.
2024/01/30 23:43:34 WARNING mlflow.metrics.metric_definitions: Failed to load automated readability index metric, skipping metric logging.
2024/01/30 23:43:34 WARNING mlflow.models.evaluation.default_evaluator: Did not log builtin metric 'ari_grade_level' because it returned None.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:05<00:00,  5.40s/it]
answer sim None
2024/01/30 23:43:39 WARNING mlflow.metrics.metric_definitions: Failed to load 'toxicity' metric (error: ModuleNotFoundError("No module named 'evaluate'")), skipping metric logging.
2024/01/30 23:43:39 WARNING mlflow.models.evaluation.default_evaluator: Did not log builtin metric 'toxicity' because it returned None.
2024/01/30 23:43:39 WARNING mlflow.metrics.metric_definitions: Failed to load flesch kincaid metric, skipping metric logging.
2024/01/30 23:43:39 WARNING mlflow.models.evaluation.default_evaluator: Did not log builtin metric 'flesch_kincaid_grade_level' because it returned None.
2024/01/30 23:43:39 WARNING mlflow.metrics.metric_definitions: Failed to load automated readability index metric, skipping metric logging.
2024/01/30 23:43:39 WARNING mlflow.models.evaluation.default_evaluator: Did not log builtin metric 'ari_grade_level' because it returned None.
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:08<00:00,  2.99s/it]
answer sim None
<mlflow.models.evaluation.base.EvaluationResult object at 0x2b21f1810>
Downloading artifacts: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 1546.00it/s]
            inputs                                       ground_truth  ... answer_similarity/v1/score                 answer_similarity/v1/justification
0  What is MLflow?  MLflow is an open-source platform for managing...  ...                          4  The model's output accurately describes MLflow...
1   What is Spark?  Apache Spark is an open-source, distributed co...  ...                          4  The model's output accurately describes Apache...
2  What is Python?  Python is a high-level programming language th...  ...                          4  The model's output accurately describes Python...

[3 rows x 6 columns]
```

Reply:
Seems like we should do two things here:
1) Don't allow users to name metrics with "/" (this conflicts with versioning and metric_name/score or metric_name/p90 as well) and perhaps warn about other symbols, currently we don't have such metric name validation.
2) Get the dependent metric based on the name of the metric (without the version) - this should've been done in this PR 😓

Wdyt? Thanks a lot for catching this and even testing it!

Reply:
Sounds great! :)